experimental/Schemes/CoveredScheme.jl

```diff
@@ -0,0 +1,839 @@
+
+    ρ : C1 → C2
+
+where for each patch ``U`` in `C1` the inclusion map ``ρ[U] : U → V`` 
+into the corresponding patch ``V`` of `C2` is an open embedding for which 
+both ``𝒪(U)`` and ``𝒪(V)`` have the same `base_ring` (so that they can be 
+canonically compared). 
+"""
+@attributes mutable struct CoveredScheme{CoveringType<:Covering, CoveringMorphismType<:CoveringMorphism}
+  coverings::Vector{CoveringType}
+  refinements::Dict{Tuple{CoveringType, CoveringType}, CoveringMorphismType}
+  refinement_graph::Graph{Directed}
+
+  default_covering::CoveringType
+
+  function CoveredScheme(coverings::Vector{CoveringType}, refinements::Dict{Tuple{CoveringType, CoveringType}, CoveringMorphismType}) where {CoveringType<:Covering, CoveringMorphismType<:CoveringMorphism}
+    # TODO: Check whether the refinements form a connected graph.
+    X = new{CoveringType, CoveringMorphismType}(coverings, refinements)
+    X.default_covering = X.coverings[1]
+    return X
+  end
+end
+
+### type getters
+covering_type(X::CoveredScheme{S, T}) where {S, T} = S
+covering_type(::Type{CoveredScheme{S, T}}) where {S, T} = S
+covering_morphism_type(X::CoveredScheme{S, T}) where {S, T} = T
+covering_morphism_type(::Type{CoveredScheme{S, T}}) where {S, T} = T
+affine_patch_type(X::CoveredSchemeType) where {CoveredSchemeType<:CoveredScheme} = affine_patch_type(covering_type(CoveredSchemeType))
+affine_patch_type(::Type{CoveredSchemeType}) where {CoveredSchemeType<:CoveredScheme} = affine_patch_type(covering_type(CoveredSchemeType))
+
+### type constructors
+covered_scheme_type(::Type{T}) where {T<:Spec} = CoveredScheme{covering_type(T), morphism_type(covering_type(T))}
+covered_scheme_type(X::Spec) = covered_scheme_type(typeof(X))
+
+covered_scheme_type(::Type{T}) where {T<:ProjectiveScheme} = covered_scheme_type(affine_patch_type(P))
+covered_scheme_type(P::ProjectiveScheme) = covered_scheme_type(typeof(P))
+
+### getter methods
+coverings(X::CoveredScheme) = X.coverings
+refinements(X::CoveredScheme) = X.refinements
+getindex(X::CoveredScheme, C::CoveringType, D::CoveringType) where {CoveringType<:Covering} = X.refinements[(C, D)]
+setindex(X::CoveredScheme, f::CoveringMorphismType, C::CoveringType, D::CoveringType) where {CoveringMorphismType<:CoveringMorphism, CoveringType<:Covering} = X.refinements[(C, D)]
+default_covering(X::CoveredScheme) = X.default_covering
+getindex(X::CoveredScheme, i::Int) = coverings(X)[i]
+
+patches(X::CoveredScheme) = patches(default_covering(X))
+glueings(X::CoveredScheme) = glueings(default_covering(X))
+
+function getindex(X::CoveredScheme, C::Covering)
+  for i in 1:length(coverings(X))
+    C == coverings(X)[i] && return i
+  end
+  error("covering not listed")
+end
+
+getindex(X::CoveredScheme, i::Int, j::Int) = X[X[i], X[j]]
+
+
+set_name!(X::CoveredScheme, name::String) = set_attribute!(X, :name, name)
+name_of(X::CoveredScheme) = get_attribute(X, :name)::String
+has_name(X::CoveredScheme) = has_attribute(X, :name)
```

Comment:
I guess these could be added to AbstractAlgebra alongside `set_name!` there (of course after changing them to be for `Any`